### PR TITLE
Faster suppressions

### PR DESCRIPTION
**Pull request by sharwell**
Fixes performance issues when using diagnostic suppressors.

### CODE REVIEW

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs [hunk 1]
```diff
@@ -48,5 +49,35 @@ public static Suppression Create(SuppressionDescriptor descriptor, Diagnostic su
...
+        }
+
+        public static bool operator !=(Suppression left, Suppression right)
+        {
+            return !(left == right);
+        }
+
+        public override bool Equals(object obj)
+        {
+            return obj is Suppression suppression
```
Comment (333fred) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:65]: No reference fast path on this one?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs [hunk 2]
```diff
@@ -48,5 +49,35 @@ public static Suppression Create(SuppressionDescriptor descriptor, Diagnostic su
...
+                && Equals(suppression);
+        }
+
+        public bool Equals(Suppression other)
+        {
+            return EqualityComparer<SuppressionDescriptor>.Default.Equals(Descriptor, other.Descriptor)
```
Comment (333fred) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:71]: Or this one?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs [hunk 3]
```diff
@@ -48,5 +49,35 @@ public static Suppression Create(SuppressionDescriptor descriptor, Diagnostic su
...
+                && EqualityComparer<Diagnostic>.Default.Equals(SuppressedDiagnostic, other.SuppressedDiagnostic);
+        }
+
+        public override int GetHashCode()
+        {
+            int hashCode = 1755072348;
+            hashCode = hashCode * -1521134295 + EqualityComparer<SuppressionDescriptor>.Default.GetHashCode(Descriptor);
```
Comment (333fred) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:78]: We have hashcode helpers for combining hashcodes.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:65]: The containing type is a value type.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs (hunk 2 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:71]: The containing type is a value type.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs (hunk 3 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs]: Switched to using that.

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for a second review. This is a dogfood blocker to fix for 16.6 Preview 3.

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for a second review.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs [hunk 4]
```diff
@@ -48,5 +50,34 @@ public static Suppression Create(SuppressionDescriptor descriptor, Diagnostic su
...
+
+        public override bool Equals(object obj)
+        {
+            return obj is Suppression suppression
+                && Equals(suppression);
+        }
+
+        public bool Equals(Suppression other)
+        {
+            return EqualityComparer<SuppressionDescriptor>.Default.Equals(Descriptor, other.Descriptor)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:72]: >EqualityComparer<SuppressionDescriptor>.Default.Equals(Descriptor, other.Descriptor) [](start = 19, length = 84)

Why go through an extra indirection, it looks like neither Descriptor, nor SuppressedDiagnostic can be null and both expose strongly typed Equals?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs [hunk 5]
```diff
@@ -48,5 +50,34 @@ public static Suppression Create(SuppressionDescriptor descriptor, Diagnostic su
...
+                && EqualityComparer<Diagnostic>.Default.Equals(SuppressedDiagnostic, other.SuppressedDiagnostic);
+        }
+
+        public override int GetHashCode()
+        {
+            return Hash.Combine(
+                EqualityComparer<SuppressionDescriptor>.Default.GetHashCode(Descriptor),
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:79]: >EqualityComparer<SuppressionDescriptor>.Default.GetHashCode(Descriptor) [](start = 16, length = 71)

Why go through an extra indirection, it looks like neither Descriptor, nor SuppressedDiagnostic can be null?

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 4)

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs (hunk 4 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:72]: The containing type is a value type. This pattern ensures the default value for the type does not throw exceptions from `Equals`.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs (hunk 5 above)
Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:79]: The containing type is a value type, so this field can be null.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:79]: >GetHashCode(Descriptor) [](start = 64, length = 23)

According to documentation, this is going to throw on a null input. Should we protect?

Comment (sharwell) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/Suppression.cs:79]: No. The documentation is wrong. This is the long-time safe use of this type. See dotnet/runtime#30998.

**Comment by sharwell** [comment]
@AlekseyTs responded to feedback

**Pull request merged**